lib/std/zon/parse.zig

```diff
@@ -0,0 +1,2832 @@
+            .optional => return self.parseOptional(T, options, node),
+
+            else => @compileError("type '" ++ @typeName(T) ++ "' is not available in ZON"),
+        }
+    }
+
+    fn parseBool(self: @This(), node: Zoir.Node.Index) !bool {
+        switch (node.get(self.zoir)) {
+            .true => return true,
+            .false => return false,
+            else => return self.failNode(node, "expected type 'bool'"),
+        }
+    }
+
+    fn parseInt(
+        self: @This(),
+        comptime T: type,
+        node: Zoir.Node.Index,
+    ) !T {
+        switch (node.get(self.zoir)) {
+            .int_literal => |int| switch (int) {
+                .small => |val| return std.math.cast(T, val) orelse
+                    self.failCannotRepresent(T, node),
+                .big => |val| return val.toInt(T) catch
+                    self.failCannotRepresent(T, node),
+            },
+            .float_literal => |val| return intFromFloatExact(T, val) orelse
+                self.failCannotRepresent(T, node),
+
+            .char_literal => |val| return std.math.cast(T, val) orelse
+                self.failCannotRepresent(T, node),
+            else => return self.failNodeFmt(node, "expected type '{s}'", .{@typeName(T)}),
+        }
+    }
+
+    fn parseFloat(
+        self: @This(),
+        comptime T: type,
+        node: Zoir.Node.Index,
+    ) !T {
+        switch (node.get(self.zoir)) {
+            .int_literal => |int| switch (int) {
+                .small => |val| return @floatFromInt(val),
+                .big => |val| return val.toFloat(T),
+            },
+            .float_literal => |val| return @floatCast(val),
+            .pos_inf => return std.math.inf(T),
+            .neg_inf => return -std.math.inf(T),
+            .nan => return std.math.nan(T),
+            .char_literal => |val| return @floatFromInt(val),
+            else => return self.failNodeFmt(node, "expected type '{s}'", .{@typeName(T)}),
+        }
+    }
+
+    fn parseEnumLiteral(
+        self: @This(),
+        comptime T: type,
+        node: Zoir.Node.Index,
+    ) !T {
+        switch (node.get(self.zoir)) {
+            .enum_literal => |string| {
+                // Create a comptime string map for the enum fields
+                const enum_fields = @typeInfo(T).@"enum".fields;
+                comptime var kvs_list: [enum_fields.len]struct { []const u8, T } = undefined;
+                inline for (enum_fields, 0..) |field, i| {
+                    kvs_list[i] = .{ field.name, @enumFromInt(field.value) };
+                }
+                const enum_tags = std.StaticStringMap(T).initComptime(kvs_list);
+
+                // Get the tag if it exists
+                return enum_tags.get(string.get(self.zoir)) orelse
+                    self.failUnexpectedField(T, node, null);
+            },
+            else => return self.failNode(node, "expected enum literal"),
+        }
+    }
+
+    fn parsePointer(
```

Comment:
Just throwing a comment in to track: remember to apply the type changes we discussed to `std.zon` too. (For instance, allowing initializing data through pointers.)